Joins with boolean column expressions, 2nd attempt

OlivierBlanvillain opened the pull request:
This PR replaces #162 with a nicer solution that is both safer than the original approach (columns keep their "source dataset" phantom type) and simpler than the `implicit CanAccess` approach.

The trick is to abuse Scala type system to model "a boolean columns coming from either table A or B" as `TypedColumn[A with B, Boolean]`. The only changes that makes this possible are on the `TypedColumn` operation which now take an extra type parameter for the second argument of binary operations and combines the phantom types as expected. From a uses perspective nothing changes:

```scala
val condition: TypedColumn[A with B, Boolean] = (df1: A).col('a) === (df2: B).col('b)
```

The implicit conversions using for the widden import broke with this change. I propose to remove this mechanism altogether and require users to explicitly cast when they convert columns from one type to another. Implicit conversion do not compose and break type inference left and right, these changes are a good illustration of that.

OlivierBlanvillain commented:
@kanterov Would you have time for a review?

kanterov commented:
@OlivierBlanvillain looking into it now! :)

iravid commented:
👍🏻 on explicit casts.

kanterov commented:
The trick with `A with B` is very neat. However, it seems that it messes up type inference outside of joins. I tried to play with implicit tricks we discussed #162, and everything seems to break inference in IntelliJ. The only thing that worked properly was:

```scala
  /** Lifts typed columns in join expressions */
  trait Join[L, R]

  object Join {
    def apply[L, R](): Join[L, R] = new Join[L, R] {}
  }

  implicit class LeftSyntax[L, U](c: TypedColumn[L, U]) {
    def left[R](scope: Join[L, R]): TypedColumn[Join[L, R], U] = new TypedColumn(c.expr)(c.uencoder)
  }

  implicit class RightSyntax[R, U](c: TypedColumn[R, U]) {
    def right[L](scope: Join[L, R]): TypedColumn[Join[L, R], U] = new TypedColumn(c.expr)(c.uencoder)
  }
```

With `TypedDataset` signatures like

```scala
  def joinInner[U](other: TypedDataset[U])(condition: Join[T, U] => TypedColumn[Join[T, U], Boolean])
```

And user code:

```
leftDs
  .joinLeft(rightDs)(s => leftDs.col('a).left(s) === rightDs.col('a).right(s))
```

As you can see, it's super explicit lifting of everything. It doesn't even work in `left` and `right` have the same name. I start to think that we are on very fragile edge with relying on an implicit inference that isn't super reliable and can easily break in corner cases.

I feel we can solve any syntactic problem with investing into two macros:

```
def expr(f: A => B): TypedColumn[A, B]
def joinExpr(f: (A, B) => C): TypedColumn[Join[A, B], C]
```
And having a macro-free way that doesn't necessarily has nice syntax, but must have nice type inference.

kanterov commented:
One more thing, it's important to distinguish column that comes from the left dataset, and from the right dataset. Consider expression:
```
ds.joinInner(ds)(s => ds.col('a).left(s) === ds.col('a).right(s))
```

vs.

```
ds.joinInner(ds)(s => ds.col('a).left(s) === ds.col('a).left(s))
```

Now Spark fixed this only for equality (see `TypedDataset#resolveSelfJoin`), rewriting left side of `===` against left dataset, and right side against the right dataset. It doesn't work for asymmetric operators like `>=`. 

In vanilla Spark API, in the case joined datasets aren't exactly the same, such ambiguity can be resolved by explicitly building expressions from a left and right datasets like:

```
val a: TypedDataset[X1[A]]
val b: TypedDataset[X1[A]]

a.joinInner(b, a.col('a) === b.col('a))
```

But this is very fragile to any transformations, for instance:

```
a.joinInner(b.filter(...), a.col('a) === b.col('a))
```

In this sense, being able to determine that column belongs to a right or left part would allow us to resolve this problem. Unfortunately, approach with `A with B` doesn't capture this information if both `A` and `B` have the same type.

OlivierBlanvillain commented:
> The trick with A with B is very neat. However, it seems that it messes up type inference outside of joins, and it doesn't seem right to trade off other uses cases for having nice joins.

Do you have some example I could play with? The test suite compiled fine after these changes, I thought that was enough ;)

> Unfortunately, approach with A with B doesn't capture this information if both A and B have the same type.

I didn't follow you here. Do we need this distinction at the typelevel? The proposed API mirrors vanilla Spark, it's possible to do `a.col('a) === b.col('a)` with this PR the exact same way it can be done in Spark. Self joins might be problematic, but I guess with could add `.left` & `.right` methods on columns to disambiguate only these cases, and leave the common cases untouched.

I feel I'm missing something, maybe an executable example would help me better understand the issue.

imarios commented:
@OlivierBlanvillain , @kanterov do you guys want to wait for this to close before cutting 0.4?

OlivierBlanvillain commented:
ping @kanterov

kanterov commented:
@OlivierBlanvillain sorry for the delay, I'm on vacation and have limited access to my machine, the problem comes from the fact, that column expressions like `df.col('a) + df.col('a)` infer to types like `TypedColumn[T with T, Int]`.

Scala compiler infers that `TypedColumn[T with T, Int]` is equivalent to `TypedColumn[T, Int]`, and all our code is compiling. While IntelliJ just breaks everything we have so far. I don't know implications of this on compilation time too.

Another problem is the way Spark resolves columns, I've prepared this code snippet to illustrate different kinds of problems https://gist.github.com/kanterov/49014d462b265eaf54353f09f0ac517a.

Please let me know what do you think.

OlivierBlanvillain commented:
I did a quick experiment and I think the `T with T` issue can be worked-around:

```scala
trait Col[T, U]

trait TA
trait TB

val t1: Col[TA, Boolean] = null
val t2: Col[TA, Boolean] = null
val t3: Col[TB, Boolean] = null

trait With[A, B] { type Out }

trait LowPrioWith {
  implicit def identity[T]: With[T, T] { type Out = T } = null
}

object With extends LowPrioWith {
  implicit def combine[A, B]: With[A, B] { type Out = A with B } = null
}

def ++[A, B, U](a: Col[A, U], b: Col[B, U])(implicit w: With[A, B]): Col[w.Out, U] = null

scala> ++(t1, t3)
res4: Col[TA with TB,Boolean] = null

scala> ++(t1, t2)
res5: Col[TA,Boolean] = null
```

Thanks for the gist, I will have have a look a closer look later!

imarios commented:
@OlivierBlanvillain is this PR still in progress? Can we merge parts of it?

OlivierBlanvillain commented:
No I don't think it's partially mergeable...

OlivierBlanvillain commented:
@kanterov I think I finally addressed your review, it only took 6 months :sweat_smile: 

The overall diff might be getting too large, so you might want to look at commits instead. Here are the two interesting ones:

-  52cd1cf992b513c73f62a890f9714c939e9e6ec1 Introduces a `With` type class to solves the A with A with ... with A blowup.
- 7cdec6be11e0c12ffe7829e51049d2952e9230d5 Add a self-join disambiguation mechanism.

imarios commented:
@OlivierBlanvillain This looks amazing ✨. Have been trying to break this with funky joins on options, collections, etc. Works great so far.

imarios commented on dataset/src/main/scala/frameless/TypedColumn.scala:
```diff
@@ -78,7 +76,7 @@ sealed class TypedColumn[T, U](
     * apache/spark
     */
-  def =!=(other: TypedColumn[T, U]): TypedColumn[T, Boolean] = withExpr {
+  def =!=[TT, W](other: TypedColumn[TT, U])(implicit w: With.Aux[T, TT, W]): TypedColumn[W, Boolean] = withExpr {
```
Can we go a bit over the `With` type class? why do we need to do this?

OlivierBlanvillain replied:
I answered in the main thread to not lose the message on rebase

imarios commented on dataset/src/main/scala/frameless/TypedDataset.scala:
```diff
@@ -203,6 +200,50 @@ class TypedDataset[T] protected[frameless](val dataset: Dataset[T])(implicit val
+  }
+
+  /** Left hand side disambiguation of `col` for join expressions.
+    * To be used  when writting self-joins, noop in other circonstances.
```
French : )? circonstances

OlivierBlanvillain replied:
:fr:

OlivierBlanvillain commented:
> Can we go a bit over the `With` type class? why do we need to do this?

So, suppose you want to write a funky join:

```
dsA.join(dsB)(dsA.col('a) != dsB.col('b))
```

Where dsA and dsB have unrelated types A and B. The signature of `join` should be such that it allows any (boolean) expression involving dsA and dsB (but not dsC). The way this is done is by having `dsA.col('a) != dsB.col('b)` returning a `TypedColumn[A with B, Boolean]`, which is precisely what the join expects. `With` is here to compute `A with B`.

You might wonder, why not just write `A with B` in the signature of `!=` in the first place? The reason is that compilers will keep track of unnecessary duplicates formed using `with`, such that expressions like  `dsA('a) + dsA('x) + ds('y)` have type `A with A with A`, something that might confuse users and compilers :smile:

OlivierBlanvillain commented:
@imarios This is the expected behavior. the join condition you wrote is trivially true, so it's equivalent to writing  `t.joinFull(t2)(lit(true))`. Spark will complain about that unless you manually set `conf.set("spark.sql.crossJoin.enabled", "true")`.

If you want to write a self join, you need to use the explicit `colLeft` and `colRight` methods:

```scala
t2.joinFull(t2)(t2.colLeft('_1) === t2.colRight('_1)).show().run
```

But there is not much we can do about trivial join condition, for example, you also explicitly write 

```scala
t2.joinFull(t2)(t2.colLeft('_1) === t2.colLeft('_1)).show().run
```

imarios commented:
@OlivierBlanvillain I see that the condition doesn't make sense, but probably this will be the only Frameless op that leads to run-time error but we say it's an expected behavior. Currently, anything that would cause a runtime exception gives compilation error, which is pretty much our sale point. The only current exception that we know of is explode(), which we already classified as a bug.

OlivierBlanvillain commented:
> Can't we just enforce that the joined condition (Boolean column) is either of type A with B or B with A, but not A with A, B with B, or A with C etc.?

It's more complicated than that, for instance expressions like `(A.a == A.a) || B.b` involves both `A` and `B` but is trivially true and probably rejected by Spark. To me this is a configuration problem: joins in Spark/frameless are only sound under the following settings:

```scala
conf.set("spark.sql.selfJoinAutoResolveAmbiguity", "true")
conf.set("spark.sql.crossJoin.enabled", "true")
```

There is no way to enforce that at compile time, so I'm not sure what's would be the best option to enforce such configuration... (we already have the first requirement, see [here](https://github.com/typelevel/frameless/blob/master/dataset/src/main/scala/frameless/TypedDataset.scala#L350-L351))

imarios commented:
Fair enough, I guess we can always try to improve on this on a seperate PR. Do you think we can do better in terms of test coverage?

OlivierBlanvillain commented:
Sure, I know what's missing. I'm just waiting for #153 to basically redo everything on `TypedColumn.scala` :smile:

imarios commented:
Hey @OlivierBlanvillain this is the last code related PR for 0.5! I know you must be busy, so whenever you have the time. Thank you!

OlivierBlanvillain commented:
@imarios It's rebased and should be hopefully be green :smile:

imarios commented:
Looks good @OlivierBlanvillain! Do you mind the code coverage concerns from Travis?

OlivierBlanvillain commented:
The two lines not covered are [these two](https://github.com/typelevel/frameless/pull/175/files#diff-fe163bbd28024dcd98ae2a5d4355cd33R61), not sure how to trigger genCode but here I simply forwards to the underlying `Expression`. Note that the `doGenCode` is only called by the original `genCode` (it cannot be called from the outside given that it's protected), so it's OK to have the `???` there.
